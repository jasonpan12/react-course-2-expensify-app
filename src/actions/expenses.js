// ADD_EXPENSE
import uuid from "uuid";

export const addExpense = ({
		description = '',
		note = '',
		amount = 0,
		createdAt = 0
} = {}) => ({
	type: 'ADD_EXPENSE',
	expense: {
		id: uuid(),
		description,
		note,
		amount,
		createdAt
	}
});

// REMOVE_EXPENSE
export const removeExpense = ({id} = {}) => ({ // destructure object passed in, and destructure empty if no
	type: 'REMOVE_EXPENSE', // this is the object that is returned, implicitly return an object (don't have to say return)
	id
});

// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
	type: 'EDIT_EXPENSE',
	id,
	updates
});
