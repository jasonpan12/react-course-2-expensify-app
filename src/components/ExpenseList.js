import React from 'react';
import {connect} from 'react-redux';
import ExpenseListItem from "./ExpenseListItem";
import selectExpenses from '../selectors/expenses';

// Export this so we can test the unconnected version in jest
export const ExpenseList = (props) => (
	<div>
		{
			props.expenses.length === 0 ? (
				<p>No expenses</p>
			) : (
				props.expenses.map((expense) => {
					return <ExpenseListItem key={expense.id} {...expense}/> // spread operator
				})
			)
		}
	</div>
);

// Takes state and makes it into props for the connect api
// basically what do you want off the store, then what component do you want to connect the store to
const mapStateToProps = (state) => ({
	// Determine which expenses we should show in the list
	// by using selectExpenses
	expenses: selectExpenses(state.expenses, state.filters)
})

// call it ConnectedExpenseList
// connect returns a function, which means we need to call that function
export default connect(mapStateToProps)(ExpenseList);
