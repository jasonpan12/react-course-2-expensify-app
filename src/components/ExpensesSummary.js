import React from 'react';
import selectExpenses from '../selectors/expenses'
import selectExpensesTotal from "../selectors/expenses-total";
import {connect} from "react-redux";
import numeral from "numeral";

// Must export this on its own (unconnected) in order to test the unconnected component
export const ExpensesSummary = (props) => {
	const expenseCount = props.expenses.length;
	const expenseWord = expenseCount === 1 ? 'expense' : 'expenses';
	const expensesTotal = numeral(selectExpensesTotal(props.expenses)/100).format('$0,0.00');
	return (
		<h1>{`Viewing ${expenseCount} expenses totalling ${expensesTotal}`}</h1>
	);
}

// Map redux state to props. props.expenses is the result of selectExpenses.
const mapStateToProps = (state) => ({
	expenses: selectExpenses(state.expenses, state.filters)
})

// Connect this component to the Redux store
export default connect(mapStateToProps)(ExpensesSummary);
