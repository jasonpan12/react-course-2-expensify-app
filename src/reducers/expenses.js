
// expenses reducer
const expensesReducerDefaultState = [];
export default (state = expensesReducerDefaultState, action) => {
	switch (action.type) {
		case 'ADD_EXPENSE':
			return [
				...state,
				action.expense
			];
		case 'REMOVE_EXPENSE':
			return state.filter(({id}) => id !== action.id); // remove item  if statement resolves to false (when id == action.id)
		case 'EDIT_EXPENSE':
			return state.map((expense) => {
				if (expense.id === action.id) {
					return {
						...expense,
						...action.updates
					};
				} else {
					return expense;
				}
			});
		default:
			return state;
	}
};

