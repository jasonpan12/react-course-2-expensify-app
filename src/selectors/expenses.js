// Get visible expenses
import moment from "moment";

export default (expenses, { text, sortBy, startDate, endDate}) => { // destructured filter object
	return expenses.filter((expense) => {
		// for every expense, see if it should be returned by filter
		const createdAtMoment = moment(expense.createdAt); // create a moment and use it below
		const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day'): true;
		const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment,'day'): true;
		// const startDateMatch = typeof startDate !== 'number' || expense.createdAt >= startDate; // default to true if startDate is not a number... fail "open"
		// const endDateMatch = typeof endDate !== 'number' || expense.createdAt <= endDate;
		const textMatch = typeof text !== 'string' || expense.description.toLowerCase().includes(text);

		// figure out if expenses.description has the text variable string inside of it
		// includes sees if one string includes another
		// convert both strings to lower case

		return startDateMatch && endDateMatch && textMatch // if all are true, filter function return true, item kept in array
		// if false, item removed from array
	}).sort((a, b) => {
		if (sortBy === 'date') {
			return a.createdAt < b.createdAt ? 1 : -1 // smaller value gets lower index in the array
		} else if (sortBy === 'amount') {
			return a.amount < b.amount ? 1 : -1
		}

		// sortBy -> amount
		// put the ones with a greater amount first
	});
}
