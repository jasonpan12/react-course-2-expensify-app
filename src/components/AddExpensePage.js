import React from "react";
import {connect} from 'react-redux';
import ExpenseForm from "./ExpenseForm";
import {startAddExpense} from "../actions/expenses";

export class AddExpensePage extends React.Component {
	onSubmit = (expense) => {
		this.props.startAddExpense(expense);
		this.props.history.push('/');
	}

	render() {
		return (
			<div>
				<h1>Add Expense</h1>
				<ExpenseForm
					onSubmit={this.onSubmit}
				/>
			</div>
		)
	}
}

const mapDispatchToProps = (dispatch) => ({
	startAddExpense: (expense) => dispatch(startAddExpense(expense)) // implicitly return an object, where onSubmit is a function
	// pull addExpense into here so that we can test this component
	// should be named addExpense, the name of the dispatcher
	// return your dispatcher functions here
})

export default connect(undefined, mapDispatchToProps)(AddExpensePage);
