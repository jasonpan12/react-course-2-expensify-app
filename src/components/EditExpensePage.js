import React from "react";
import {connect} from "react-redux";
import ExpenseForm from "./ExpenseForm";
import {startEditExpense, startRemoveExpense} from "../actions/expenses";

// Refactor EditExpensePage to be a class based component, which will let us pull out the inline fnctions
// set up mapDispatchToProps editExpense and removeExpense

// should render EditExpensePage
// snapshot

// should handle editExpense
// spies

// should handle removeExpense
// spies

export class EditExpensePage extends React.Component {
	onSubmit = (expense) => { // by binding the onSubmit here, we aren't re-rendering onsubmit each time the props update
		// the expense object is provided down in ExpenseForm
		// dispatch action to edit expense
		// redirect to dashboard page
		this.props.startEditExpense(this.props.expense.id, expense)
		this.props.history.push('/');
	}

	onRemove = () => {
		// dispatch action to removeExpense
		this.props.startRemoveExpense({id: this.props.expense.id})
		this.props.history.push('/');
	}

	render() {
		return (
			<div>
				<div className = "page-header">
					<div className = "content-container">
						<h1 className = "page-header__title">Edit Expense</h1>
					</div>
				</div>
				<div className = "content-container">
					<ExpenseForm
						expense = {this.props.expense}
						onSubmit = {this.onSubmit}
					/>
					<div>
						<button className = "button button--secondary" onClick={this.onRemove}>Remove Expense</button>
					</div>
				</div>
			</div>
		)
	}
}

const mapStateToProps = (state, props) => { // these props can be sent down to the HOC
	return { // EditExpensePage is given an expense prop
		expense: state.expenses.find((expense) => expense.id === props.match.params.id)
		// All expenses are in the state, but this page has an id in its params
		// the expense prop is whichever expense matches the ID in the params.
	}
}

const mapDispatchToProps = (dispatch, props) => ({
	startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
	startRemoveExpense: (data) => dispatch(startRemoveExpense(data))
})

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage); // this is the HOC (EditExpensePage), and we can pass new props in
