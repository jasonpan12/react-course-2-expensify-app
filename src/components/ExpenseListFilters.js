import React from 'react';
import {connect} from 'react-redux';
import {DateRangePicker} from 'react-dates';
import {setTextFilter, sortByAmount, sortByDate, setStartDate, setEndDate} from "../actions/filters";

export class ExpenseListFilters extends React.Component {
	state = {
		calendarFocused: null // ExpenseForm  also kept track of calendarFocused
	}
	onDatesChange = ({startDate, endDate}) => { // onDatesChange will be called by the react-dates library with an object startDate and endDate
		this.props.setStartDate(startDate);
		this.props.setEndDate(endDate);
	}
	onFocusChange = (calendarFocused) => {
		this.setState(() => ({calendarFocused}))
	}

	onTextChange = (event) => {
		// dispatch gets included in props with connected components
		this.props.setTextFilter(event.target.value);
	}

	onSortChange = (event) => {
		if (event.target.value === "date") {
			this.props.sortByDate()
		} else if (event.target.value === "amount") {
			this.props.sortByAmount()
		}
	}

	render() {
		return (
			<div>
				<input placeholder="Text Filter" type="text" value={this.props.filters.text}
					   onChange = {this.onTextChange}/>
				<select value={this.props.filters.sortBy} onChange = {this.onSortChange}>
					<option value="date">Date</option>
					<option value="amount">Amount</option>
				</select>
				<DateRangePicker
					startDate={this.props.filters.startDate}
					endDate={this.props.filters.endDate}
					onDatesChange={this.onDatesChange}
					focusedInput={this.state.calendarFocused}
					onFocusChange={this.onFocusChange}
					showClearDates={true}
					numberOfMonths={1}
					isOutsideRange={() => false}
				/>
			</div>
		);
	}
}

// Pulls state from redux store, passes in filters as props
const mapStateToProps = (state) => {
	return {
		filters: state.filters
	};
}

const mapDispatchToProps = (dispatch) => ({ // factor all dispatch calls down here
	setTextFilter: (text) => dispatch(setTextFilter(text)),
	sortByDate: () => dispatch(sortByDate()),
	sortByAmount: () => dispatch(sortByAmount()),
	setStartDate: (startDate) =>  dispatch(setStartDate(startDate)),
	setEndDate: (endDate) => dispatch(setEndDate(endDate))
})
// Map dispatch functions to Props
export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);
