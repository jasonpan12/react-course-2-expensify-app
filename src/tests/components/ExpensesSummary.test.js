import {shallow} from "enzyme/build";
import expenses from "../fixtures/expenses";
import React from "react";
import {ExpensesSummary} from "../../components/ExpensesSummary";

test('should display correct summary for one expense', () => {
	const wrapper = shallow(<ExpensesSummary expenses={[expenses[0]]}/>);
	expect(wrapper).toMatchSnapshot();
})

test('should display correct summary for two expenses', () => {
	const wrapper = shallow(<ExpensesSummary expenses={[expenses[0], expenses[1]]}/>);
	expect(wrapper).toMatchSnapshot();
})
