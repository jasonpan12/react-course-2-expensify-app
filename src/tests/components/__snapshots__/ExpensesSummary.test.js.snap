// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should display correct summary for one expense 1`] = `
<h1>
  Viewing 1 expenses totalling $1.95
</h1>
`;

exports[`should display correct summary for two expenses 1`] = `
<h1>
  Viewing 2 expenses totalling $1,096.95
</h1>
`;
